pkg/segment/writer/metrics/metricssegment.go

```diff
@@ -1754,27 +1751,32 @@ func FindTagValuesUnrotated(tRange *dtu.MetricsTimeRange, mQuery *structs.Metric
 	return nil
 }
 
-func (mb *MetricsBlock) appendToWALBuffer(dp wal.WalDatapoint) error {
-	mb.walDPBuffer = append(mb.walDPBuffer, dp)
-	if len(mb.walDPBuffer) > utils.MAX_WALDATPOINT_SIZE_TO_FLUSH {
-		err := mb.currentWal.AppendToWAL(mb.walDPBuffer)
+func (mb *MetricsBlock) appendToWALBuffer(timestamp uint64, dp float64, tsid uint64) error {
+	mb.lock.Lock()
+	defer mb.lock.Unlock()
+	mb.dpsInWalMem[mb.dpIdx].Timestamp = timestamp
+	mb.dpsInWalMem[mb.dpIdx].DpVal = dp
+	mb.dpsInWalMem[mb.dpIdx].Tsid = tsid
+	mb.dpIdx++
+	if len(mb.dpsInWalMem) >= utils.WAL_BLOCK_FLUSH_SIZE {
+		err := mb.currentWal.AppendToWAL(mb.dpsInWalMem)
```

Comment:
don;t you need to do `mb.dpsInWalMem[0:mb.dpIdx]` here to pick the correct amount of DPs